cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -0,0 +1,280 @@
+                                    Class<?> keyClass, Class<?> valueClass, Map<String, String> outputConfigs) {
+    checkNamedOutputName(namedOutput, getNamedOutputsList(job), false);
+    Configuration conf = job.getConfiguration();
+    conf.set(MULTIPLE_OUTPUTS, conf.get(MULTIPLE_OUTPUTS, "") + " " + namedOutput);
+    conf.set(MO_PREFIX + namedOutput + FORMAT, outputFormatClass);
+    conf.setClass(MO_PREFIX + namedOutput + KEY, keyClass, Object.class);
+    conf.setClass(MO_PREFIX + namedOutput + VALUE, valueClass, Object.class);
+    setNamedConfigurations(job, namedOutput, outputConfigs);
+  }
+
+  // instance code, to be used from Mapper/Reducer code
+
+  private TaskInputOutputContext context;
+  private Set<String> namedOutputs;
+  private Map<String, RecordWriter<?, ?>> recordWriters;
+
+  /**
+   * Creates and initializes multiple outputs support,
+   * it should be instantiated in the Mapper/Reducer setup method.
+   *
+   * @param context the TaskInputOutputContext object
+   */
+  public MultipleOutputs(TaskInputOutputContext context) {
+    this.context = context;
+    namedOutputs = Collections.unmodifiableSet(
+      new HashSet<>(MultipleOutputs.getNamedOutputsList(context)));
+    recordWriters = new HashMap<>();
+  }
+
+  /**
+   * Write key and value to the namedOutput.
+   *
+   * @param namedOutput the named output name
+   * @param key         the key
+   * @param value       the value
+   */
+  @SuppressWarnings("unchecked")
+  public <K, V> void write(String namedOutput, K key, V value) throws IOException, InterruptedException {
+    checkNamedOutputName(namedOutput, namedOutputs, true);
+    getRecordWriter(namedOutput).write(key, value);
+  }
+
+  // by being synchronized MultipleOutputTask can be use with a MultithreadedMapper.
+  @SuppressWarnings("unchecked")
+  private synchronized RecordWriter getRecordWriter(String namedOutput) throws IOException, InterruptedException {
+    // look for record-writer in the cache
+    RecordWriter writer = recordWriters.get(namedOutput);
+
+    // If not in cache, create a new one
+    if (writer == null) {
+      // get the record writer from context output format
+      try {
+        TaskAttemptContext taskContext = getContext(namedOutput);
+
+        Instantiator<? extends OutputFormat<?, ?>> instantiator =
+          new InstantiatorFactory(false).get(TypeToken.of(taskContext.getOutputFormatClass()));
+        OutputFormat<?, ?> outputFormat = instantiator.create();
+
+        writer = outputFormat.getRecordWriter(taskContext);
+      } catch (ClassNotFoundException e) {
+        throw new IOException(e);
+      }
+
+      // add the record-writer to the cache
+      recordWriters.put(namedOutput, writer);
+    }
+    return writer;
+  }
+
+  // Create a taskAttemptContext for the named output with
+  // output format and output key/value types put in the context
+  private TaskAttemptContext getContext(String nameOutput) throws IOException {
```

Comment:
Why this method doesn't need to synchornized while the getRecordWriter needs to? Both of them seems to be looking up and modifying cache.